config/services.yaml

```diff
@@ -68,15 +81,17 @@ services:
 
     Atoolo\GraphQL\Search\Resolver\ArticleTeaserResolver:
         arguments:
-          - '@Atoolo\GraphQL\Search\Resolver\ResourceAssetResolver'
+          - '@Atoolo\GraphQL\Search\Resolver\Asset\ResourceAssetResolverChain'
+          - '@Atoolo\GraphQL\Search\Resolver\Asset\ResourceAssetResolverChain'
```

Comment:
Here is a better one:
```yml 
- '@Atoolo\GraphQL\Search\Resolver\Asset\ResourceAssetResolverChain' 
- '@Atoolo\GraphQL\Search\Resolver\Asset\ResourceSymbolicImageHierachyResolver' 
```

Reply:
The second argument (`$symbolicImageResolver`) also needs to be a chain (and optimally the same chain), because customers might want to have specialized resolving mechanisms for symbolic images that they can then register as custom services. (e.g. certain object types always resolving to a certain symbolic image)

Maybe i should further explain my general idea. Let's say we have a chain of (made-up) asset resolvers, some of which are also symbolic images resolvers:
1. PrimaryImageResolver (implements `ResourceAssetResolver`)
2. SomeOtherAssetResolver (implements `ResourceAssetResolver`)
3. SymbolicImageByObjectTypeResolver (implements `ResourceAssetResolver` & `ResourceSymbolicImageResolver`)
4. SymbolicImageHierarchyResolver (implements `ResourceAssetResolver` & `ResourceSymbolicImageResolver`)
5. LastImageResolver (implements `ResourceAssetResolver`)

So when the client requests the asset field via
```gql
teaser {
  ...
  asset(variant: "teaser")
}
```
the teaser resolver is called, which calls the `getAsset` method, which calls the chain that iterates over every resolver 1,2,3,4 and 5 until it gives a non-null result. This is the "general purpose" case for resolving assets.

When the client requests only the symbolic image via
```gql
teaser {
  ...
  symbolicImage
}
```
the teaser resolver is called, which calls the `getSymbolicImage` method, which calls the chain that this time only iterates over resolver 3 and 4, because only they are symbolic image resolvers.

Reply:
> The second argument ($symbolicImageResolver) also needs to be a chain (and optimally the same chain), because customers might want to have specialized resolving mechanisms for symbolic images that they can then register as custom services. (e.g. certain object types always resolving to a certain symbolic image)

Yes, the only thing that bothers me is that `ResourceAssetResolverChain` has two tasks.

There should be a separate resolver for the use case that I only want to have one SymbolicImage. Currently we don't need a chain here either. And we may never need one. It is unnecessarily complex.

Reply:
Fair enough, I simplified the code such that `ResourceAssetResolverChain` now only has a single task. 

The `ResourceSymbolicImageHierarchyResolver` is now passed directly to the `ArticleTeaserResolver`/`NewsTeaserResolver` as you suggested. It can still be overriden in customer projects anyway, so I agree that my idea was a little bit overengineered. :smile: